src/Sema.zig

```diff
@@ -5512,16 +5512,156 @@ fn zirArrayCat(sema: *Sema, block: *Scope.Block, inst: Zir.Inst.Index) CompileEr
+            if (rhs_val.tag() == .decl_ref and lhs_val.tag() == .decl_ref) {
+                var anon_decl = try block.startAnonDecl();
+                defer anon_decl.deinit();
+
+                const lhs_sub_val = lhs_val.castTag(.decl_ref).?.data.val;
+                const rhs_sub_val = rhs_val.castTag(.decl_ref).?.data.val;
+                if (rhs_sub_val.tag() == .empty_array) return lhs;
+                if (lhs_sub_val.tag() == .empty_array) return rhs;
+                const buf = try anon_decl.arena().alloc(Value, final_len);
+                {
+                    var i: u64 = 0;
+                    while (i < lhs_info.len) : (i += 1) {
+                        const val = try lhs_sub_val.elemValue(sema.arena, i);
+                        buf[i] = try val.copy(anon_decl.arena());
+                    }
+                }
+                {
+                    var i: u64 = 0;
+                    while (i < rhs_info.len) : (i += 1) {
+                        const val = try rhs_sub_val.elemValue(sema.arena, i);
+                        buf[lhs_info.len + i] = try val.copy(anon_decl.arena());
+                    }
+                }
+                const ty = if (res_sent) |rs|
+                    try Type.Tag.array_sentinel.create(anon_decl.arena(), .{ .len = final_len, .elem_type = lhs_info.array_elem_type, .sentinel = rs })
+                else
+                    try Type.Tag.array.create(anon_decl.arena(), .{ .len = final_len, .elem_type = lhs_info.array_elem_type });
+                const val = try Value.Tag.array.create(anon_decl.arena(), buf);
+                return sema.analyzeDeclRef(try anon_decl.finish(
+                    ty,
+                    val,
+                ));
+            }
+            return sema.mod.fail(&block.base, lhs_src, "TODO array_cat more types of Values", .{});
+        } else {
+            return sema.mod.fail(&block.base, lhs_src, "TODO runtime array_cat", .{});
+        }
+    } else {
+        return sema.mod.fail(&block.base, lhs_src, "TODO runtime array_cat", .{});
+    }
+}
+
+const ArrayCatInfo = struct { array_elem_type: Type, sent: ?Value = null, len: u64 };
+fn getArrayCatInfo(t: Type) ?ArrayCatInfo {
+    var info: ArrayCatInfo = .{ .array_elem_type = undefined, .len = undefined };
+    if (t.zigTypeTag() == .Array) {
+        info.len = t.arrayLen();
+        if (t.tag() == .array_u8) {
+            info.array_elem_type = Type.initTag(.u8);
+        } else if (t.tag() == .array_u8_sentinel_0) {
+            info.array_elem_type = Type.initTag(.u8);
+            info.sent = Value.initTag(.zero);
+        } else if (t.tag() == .array) {
+            const payload = t.castTag(.array).?.data;
+            info.array_elem_type = payload.elem_type;
+        } else if (t.tag() == .array_sentinel) {
+            const payload = t.castTag(.array_sentinel).?.data;
+            info.array_elem_type = payload.elem_type;
+            info.sent = payload.sentinel;
+        }
```

Comment:
Should there be an error case here, similar to the case where it is neither an array nor a pointer?
```suggestion
        } else return null;
```

Reply:
The correct thing here is else unreachable; since no other type tags will say the zigTypeTag is .Array.